Space application supporter can get, create and delete service route bindings

will-gant opened the pull request:
* Enable space application supporter to get, create and delete service route bindings
* Add unit tests to prove the above

Note that the user story related to this PR did not ask for the DELETE endpoint to be implemented, but [I have assumed](https://github.com/cloudfoundry/cloud_controller_ng/issues/2242#issuecomment-874092400) that this was a mistake. If that turns out to be incorrect, I can reverse that part of the changes.

Closes #2242 

* [x] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have run all the unit tests using `bundle exec rake`

* [ ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats)

will-gant commented:
Just made a small commit to switch to using the new name of the role, and also updated the docs - something I had previously forgotten to do.

monamohebbi commented:
As per our conversation in the issue linked to this PR, could you also allow for the space supporter to use the DELETE /v3/service_route_bindings/:guid endpoint?

will-gant commented:
@monamohebbi Happily I assumed this would be your answer and made that change already [as part of this commit](https://github.com/cloudfoundry/cloud_controller_ng/pull/2369/commits/f2d839899d7dda3132509f109449861fe7c432e8).

I did, however, forget to include documentation for that so I've just updated that and squashed the change into my commit for the docs (this necessitated a force push).

sweinstein22 reviewed (commented):
## Acceptance Check

### Behavior Check

#### User Provided Service
##### Setup

Used the [Rate Limiting Route Service](https://github.com/cloudfoundry-samples/ratelimit-service)

##### `POST /v3/service_route_bindings`

Ran `cf curl` as both a space supporter and space developer, confirmed that the only diffs were guids and timestamps:
```
$ diff space-supporter-create-route-binding space-dev-create-route-binding
2c2
<   "guid": "cb6c1527-b602-4b27-8385-9e5f32c13804",
---
>   "guid": "4a5187c0-7cb5-48ca-a685-7820f6b15330",
4,5c4,5
<   "created_at": "2021-07-15T22:51:21Z",
<   "updated_at": "2021-07-15T22:51:21Z",
---
>   "created_at": "2021-07-15T22:51:59Z",
>   "updated_at": "2021-07-15T22:51:59Z",
10,11c10,11
<     "updated_at": "2021-07-15T22:51:21Z",
<     "created_at": "2021-07-15T22:51:21Z"
---
>     "updated_at": "2021-07-15T22:51:59Z",
>     "created_at": "2021-07-15T22:51:59Z"
35c35
<       "href": "https://api.clover-horse.capi.land/v3/service_route_bindings/cb6c1527-b602-4b27-8385-9e5f32c13804"
---
>       "href": "https://api.clover-horse.capi.land/v3/service_route_bindings/4a5187c0-7cb5-48ca-a685-7820f6b15330"
```

Confirmed that service plan was present in list when running `cf curl /v3/service_route_bindings` as a space developer after creating the binding as a space supporter

##### `GET /v3/service_route_bindings/:guid`

Ran `cf curl` as both a space supporter and a space developer, confirmed that ouput was the same:

```
$ diff space-supporter-route-binding-guid space-dev-route-binding-guid

```

##### `DELETE /v3/service_route_bindings/:guid`

Ran `cf curl` as both a space supporter and a space developer, confirmed that output was the same:
```
$ diff space-supporter-delete-route-binding space-dev-delete-route-binding

```

Confirmed that service plan was no longer present in list when running `cf curl /v3/service_route_bindings` as a space developer after creating the binding as a space supporter

#### Managed Service Instance

##### Setup

Pushed the [CATS Async Service Broker](https://github.com/cloudfoundry/cf-acceptance-tests/tree/develop/assets/service_broker)

Checked against plans for all visibilities:

##### Publicly visible service plan

Confirmed plan was visible at public level:

```
$ cf curl /v3/service_plans/cff73756-8fdf-4b94-9290-dcb3b27b6e26/visibility
{
  "type": "public"
}
```

Confirmed diff was equivalent for all three endpoints:
```
$ diff space-supporter-public-service-route-binding-create space-dev-public-service-route-binding-create

$ diff space-supporter-public-service-route-binding-get space-dev-public-service-route-binding-get

$ diff space-supporter-public-service-route-binding-delete space-dev-public-service-route-binding-delete

```

Checked output manually as well, the `GET` and `DELETE` outputs are what I would expect.

Interesting thing of note, when using a managed service instance the `POST` call does not return any output, however the binding is successfully created. This does not appear to be newly introduced behavior, I ran checks against the `main` branch of CAPI, and is outside the scope of this PR, but is notable.

##### Org level visible service plan

Confirmed plan was visible at org level:

```
$ cf curl /v3/service_plans/cff73756-8fdf-4b94-9290-dcb3b27b6e26/visibility
{
  "type": "organization",
  "organizations": [
    {
      "guid": "a5f5be45-bfb9-4c9a-bb2a-3bc49eeec51b",
      "name": "org"
    }
  ]
}
```

Confirmed diff was equivalent for all three endpoints:
```
$ diff space-supporter-org-level-service-route-binding-create space-dev-org-level-service-route-binding-create

$ diff space-supporter-org-level-service-route-binding-get space-dev-org-level-service-route-binding-get

$ diff space-supporter-org-level-service-route-binding-delete space-dev-org-level-service-route-binding-delete

```

Checked output manually as well, the `GET` and `DELETE` outputs are what I would expect. `POST` output is same as mentioned in public section above.

##### Org level visible service plan

Confirmed plan was visible at org level:

```
$ cf curl /v3/service_plans/cff73756-8fdf-4b94-9290-dcb3b27b6e26/visibility
{
  "type": "organization",
  "organizations": [
    {
      "guid": "a5f5be45-bfb9-4c9a-bb2a-3bc49eeec51b",
      "name": "org"
    }
  ]
}
```

Confirmed diff was equivalent for all three endpoints:
```
$ diff space-supporter-org-level-service-route-binding-create space-dev-org-level-service-route-binding-create

$ diff space-supporter-org-level-service-route-binding-get space-dev-org-level-service-route-binding-get

$ diff space-supporter-org-level-service-route-binding-delete space-dev-org-level-service-route-binding-delete

```

Checked output manually as well, the `GET` and `DELETE` outputs are what I would expect. `POST` output is same as mentioned in public section above.

##### Admin level visible service plan

Not applicable, as plan visibility is out of scope for role

### Code Check

Request spec shared examples are being used, following code patterns we've established for this work. Implementation looks good!

### Docs Check

Confirmed docs render correctly, and that new role is indicated as experimental for all three endpoints

sweinstein22 reviewed (changes requested):
### Code Check

We ended up downgrading `fog-aws` due to some issues when upgrading to `3.x.x`, if we could roll back those Gemfile changes on this branch that would be great, thanks!

### Misc Notes

Looks like we missed `GET /v3/service_route_bindings` when making the associated issue, if you feel like addressing that endpoint as well when making changes feel free, otherwise I'll open a new issue for it

iaftab-alam commented:
hey @sweinstein22,

We have roll-back those Gemfile changes. Please have a look at it and share you thoughts on this. 
For  GET /v3/service_route_bindings endpoint I would recommend to create a new issue on it.

Regards,
Aftab

sweinstein22 reviewed (approved):
Great, thanks @iaftab-alam . I've opened a new issue for `GET /v3/service_route_bindings` here: https://github.com/cloudfoundry/cloud_controller_ng/issues/2422

iaftab-alam commented:
@sweinstein22  Thanks.
